File: app/src/main/java/org/stepic/droid/ui/fragments/ProfileFragment.java
package org.stepic.droid.ui.fragments;

import android.annotation.SuppressLint;
import android.app.Activity;
import android.content.Intent;
import android.graphics.drawable.Drawable;
import android.graphics.drawable.PictureDrawable;
import android.net.Uri;
import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.v4.app.DialogFragment;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.Toolbar;
import android.view.LayoutInflater;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;
import android.widget.CompoundButton;
import android.widget.ImageView;
import android.widget.TextView;

import com.bumptech.glide.GenericRequestBuilder;
import com.bumptech.glide.Glide;
import com.bumptech.glide.load.engine.DiskCacheStrategy;
import com.caverock.androidsvg.SVG;

import org.jetbrains.annotations.NotNull;
import org.joda.time.DateTime;
import org.joda.time.DateTimeZone;
import org.joda.time.format.DateTimeFormatter;
import org.joda.time.format.DateTimeFormatterBuilder;
import org.stepic.droid.R;
import org.stepic.droid.analytic.Analytic;
import org.stepic.droid.base.App;
import org.stepic.droid.base.FragmentBase;
import org.stepic.droid.core.ProfilePresenter;
import org.stepic.droid.core.presenters.StreakPresenter;
import org.stepic.droid.core.presenters.contracts.NotificationTimeView;
import org.stepic.droid.core.presenters.contracts.ProfileView;
import org.stepic.droid.model.UserViewModel;
import org.stepic.droid.ui.custom.BetterSwitch;
import org.stepic.droid.ui.dialogs.TimeIntervalPickerDialogFragment;
import org.stepic.droid.ui.util.TimeIntervalUtil;
import org.stepic.droid.util.AppConstants;
import org.stepic.droid.util.svg.GlideSvgRequestFactory;

import java.io.InputStream;

import javax.inject.Inject;

import butterknife.BindDrawable;
import butterknife.BindString;
import butterknife.BindView;

public class ProfileFragment extends FragmentBase implements ProfileView, NotificationTimeView {

    private static final String USER_ID_KEY = "user_id_key";
    private static final int NOTIFICATION_INTERVAL_REQUEST_CODE = 11;

    public static ProfileFragment newInstance(long userId) {
        Bundle args = new Bundle();
        args.putLong(USER_ID_KEY, userId);
        ProfileFragment fragment = new ProfileFragment();
        fragment.setArguments(args);
        return fragment;
    }

    @BindView(R.id.mainInfoRoot)
    View mainInfoRoot;

    @BindView(R.id.toolbar)
    Toolbar toolbar;

    @BindView(R.id.profileName)
    TextView profileName;

    @BindView(R.id.profileImage)
    ImageView profileImage;

    @BindDrawable(R.drawable.general_placeholder)
    Drawable userPlaceholder;

    @BindView(R.id.shortBioValue)
    TextView shortBioValue;

    @BindView(R.id.aboutMeRoot)
    View aboutMeRoot;

    @BindView(R.id.currentStreakValue)
    TextView currentStreakValue;

//    @BindView(R.id.currentStreakSuffix)
//    TextView currentStreakSuffix;

//    @BindView(R.id.currentStreakPrefix)
//    TextView currentStreakPrefix;

//    @BindView(R.id.maxStreakPrefix)
//    TextView maxStreakPrefix;

    @BindView(R.id.maxStreakValue)
    TextView maxStreakValue;

//    @BindView(R.id.maxStreakSuffix)
//    TextView maxStreakSuffix;

    @BindView(R.id.shortBioTitle)
    TextView shortBioTitle;

    @BindView(R.id.infoTitle)
    TextView infoTitle;

    @BindString(R.string.about_me)
    String aboutMeTitle;

    @BindString(R.string.short_bio)
    String shortBioTitleString;

    @BindView(R.id.infoValue)
    TextView infoValue;

    @Inject
    ProfilePresenter profilePresenter;

    @Inject
    StreakPresenter streakPresenter;

    @BindView(R.id.empty_users)
    View emptyUsers;

    @BindView(R.id.loadProgressbar)
    View loadingView;

    @BindView(R.id.reportProblem)
    View reportProblemRoot;

    @BindView(R.id.content_root)
    View contentRoot;

//    @BindView(R.id.notificationStreakSwitch)
//    BetterSwitch notificationStreakSwitch;
//
//    @BindView(R.id.notificationIntervalTitle)
//    View notificationIntervalTitle;
//
//    @BindView(R.id.notificationIntervalValue)
//    TextView notificationIntervalValue;
//
//    @BindView(R.id.notificationTimeZoneInfo)
//    TextView notificationTimeZoneInfo;

    long userId;
    private UserViewModel localUserViewModel = null;

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setRetainInstance(true);
        userId = getArguments().getLong(USER_ID_KEY);
        analytic.reportEvent(Analytic.Profile.OPEN_SCREEN_OVERALL);
        setHasOptionsMenu(true);
    }

    @Override
    protected void injectComponent() {
        App.Companion
                .component()
                .profileComponentBuilder()
                .build()
                .inject(this);
    }

    @Nullable
    @Override
    public View onCreateView(LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        return inflater.inflate(R.layout.fragment_profile_new, container, false);
    }

    @Override
    public void onViewCreated(View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        initToolbar();
        initTimezone();

        profilePresenter.attachView(this);
        streakPresenter.attachView(this);
        profilePresenter.initProfile(userId);
        profileImage.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                analytic.reportEvent(Analytic.Profile.CLICK_IMAGE);
            }
        });
        View.OnClickListener clickStreakValue = new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                analytic.reportEvent(Analytic.Profile.CLICK_STREAK_VALUE);
            }
        };
        currentStreakValue.setOnClickListener(clickStreakValue);
        maxStreakValue.setOnClickListener(clickStreakValue);
        profileName.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                analytic.reportEvent(Analytic.Profile.CLICK_FULL_NAME);
            }
        });

//        notificationIntervalValue.setOnClickListener(new View.OnClickListener() {
//            @Override
//            public void onClick(View v) {
//                analytic.reportEvent(Analytic.Interaction.CLICK_CHOOSE_NOTIFICATION_INTERVAL);
//                DialogFragment dialogFragment = TimeIntervalPickerDialogFragment.Companion.newInstance();
//                if (!dialogFragment.isAdded()) {
//                    dialogFragment.setTargetFragment(ProfileFragment.this, NOTIFICATION_INTERVAL_REQUEST_CODE);
//                    dialogFragment.show(getFragmentManager(), null);
//                }
//            }
//        });
    }

    private void initTimezone() {
        DateTimeFormatter dateTimeFormatter = new DateTimeFormatterBuilder()
                .appendHourOfDay(2)
                .appendLiteral(":00 (")
                .appendTimeZoneName()
                .appendLiteral(')')
                .toFormatter();
        DateTime utc = DateTime.now(DateTimeZone.UTC).withMillisOfDay(0);
        String print = dateTimeFormatter.print(utc.withZone(DateTimeZone.getDefault()));
//        notificationTimeZoneInfo.setText(getString(R.string.streak_updated_timezone, print));
    }

    @Override
    public void onDestroyView() {
//        notificationStreakSwitch.setOnCheckedChangeListener(null);
        profileName.setOnClickListener(null);
        currentStreakValue.setOnClickListener(null);
        maxStreakValue.setOnClickListener(null);
        profileImage.setOnClickListener(null);
//        notificationIntervalValue.setOnClickListener(null);
        streakPresenter.detachView(this);
        profilePresenter.detachView(this);
        super.onDestroyView();
    }

    private void initToolbar() {
        AppCompatActivity activity = (AppCompatActivity) getActivity();
        activity.setSupportActionBar(toolbar);
        //noinspection ConstantConditions
        activity.getSupportActionBar().setDisplayHomeAsUpEnabled(true);
        activity.getSupportActionBar().setHomeAsUpIndicator(R.drawable.ic_close_white_24dp);
        activity.getSupportActionBar().setElevation(0f);
    }

    @SuppressLint("SetTextI18n")
    @Override
    public void streaksAreLoaded(int currentStreak, int maxStreak) {
        String suffixCurrent = getResources().getQuantityString(R.plurals.day_number, currentStreak);
        String suffixMax = getResources().getQuantityString(R.plurals.day_number, maxStreak);

//        currentStreakSuffix.setText(suffixCurrent);
//        maxStreakSuffix.setText(suffixMax);

        currentStreakValue.setText(currentStreak + " " + suffixCurrent);
        maxStreakValue.setText(maxStreak + " " + suffixMax);

        showStreakRoot(true);
    }

    private void showStreakRoot(boolean needShow) {
        if (needShow) {
//            currentStreakSuffix.setVisibility(View.VISIBLE);
            currentStreakValue.setVisibility(View.VISIBLE);
//            currentStreakPrefix.setVisibility(View.VISIBLE);

//            maxStreakSuffix.setVisibility(View.VISIBLE);
            maxStreakValue.setVisibility(View.VISIBLE);
//            maxStreakPrefix.setVisibility(View.VISIBLE);
        }
    }

    @Override
    public void showLoadingAll() {
        contentRoot.setVisibility(View.GONE);
        emptyUsers.setVisibility(View.GONE);
        reportProblemRoot.setVisibility(View.GONE);
        loadingView.setVisibility(View.VISIBLE);
    }

    @Override
    public void showNameImageShortBio(@NotNull UserViewModel userViewModel) {
        emptyUsers.setVisibility(View.GONE);
        reportProblemRoot.setVisibility(View.GONE);
        loadingView.setVisibility(View.GONE);
        contentRoot.setVisibility(View.VISIBLE);

        localUserViewModel = userViewModel;
        getActivity().supportInvalidateOptionsMenu();

        if (userViewModel.isMyProfile()) {
            shortBioTitle.setText(aboutMeTitle);
            streakPresenter.tryShowNotificationSetting();
        } else {
            shortBioTitle.setText(shortBioTitleString);
        }

        mainInfoRoot.setVisibility(View.VISIBLE);
        String[] nameArray = userViewModel.getFullName().split("\\s+");
        StringBuilder builder = new StringBuilder();
        for (String nameArrayItem : nameArray) {
            if (builder.length() != 0) {
                builder.append("\n");
            }
            builder.append(nameArrayItem);
        }


        profileName.setText(builder.toString());
        if (userViewModel.getImageLink() != null && userViewModel.getImageLink().endsWith(AppConstants.SVG_EXTENSION)) {
            GenericRequestBuilder<Uri, InputStream, SVG, PictureDrawable> svgRequestBuilder = GlideSvgRequestFactory.create(getContext(), userPlaceholder);
            Uri uri = Uri.parse(userViewModel.getImageLink());
            svgRequestBuilder
                    .diskCacheStrategy(DiskCacheStrategy.SOURCE)
                    .load(uri)
                    .into(profileImage);

        } else {
            Glide.with(getContext())
                    .load(userViewModel.getImageLink())
                    .asBitmap()
                    .placeholder(userPlaceholder)
                    .into(profileImage);
        }

        if (userViewModel.getShortBio().isEmpty() && userViewModel.getInformation().isEmpty()) {
            aboutMeRoot.setVisibility(View.GONE);
        } else {
            if (!userViewModel.getShortBio().isEmpty()) {
                shortBioValue.setText(userViewModel.getShortBio());
                aboutMeRoot.setVisibility(View.VISIBLE);
            } else {
                shortBioTitle.setVisibility(View.GONE);
                shortBioValue.setVisibility(View.GONE);
            }

            if (!userViewModel.getInformation().isEmpty()) {
                infoValue.setText(userViewModel.getInformation());
                aboutMeRoot.setVisibility(View.VISIBLE);
            } else {
                infoValue.setVisibility(View.GONE);
                infoTitle.setVisibility(View.GONE);
            }
        }
    }

    @Override
    public void onInternetFailed() {
        loadingView.setVisibility(View.GONE);
        contentRoot.setVisibility(View.GONE);
        emptyUsers.setVisibility(View.GONE);
        reportProblemRoot.setVisibility(View.VISIBLE);
    }

    @Override
    public void onProfileNotFound() {
        loadingView.setVisibility(View.GONE);
        contentRoot.setVisibility(View.GONE);
        reportProblemRoot.setVisibility(View.GONE);
        emptyUsers.setVisibility(View.VISIBLE);
    }

    @Override
    public void showNotificationEnabledState(boolean notificationEnabled, @NotNull String notificationTimeValueString) {
//        notificationStreakSwitch.setChecked(notificationEnabled);
//        if (notificationStreakSwitch.getVisibility() != View.VISIBLE) {
//            notificationStreakSwitch.setVisibility(View.VISIBLE);
//        }
//        if (notificationEnabled) {
//            hideNotificationTime(false);
//        } else {
//            hideNotificationTime(true);
//        }
//
//        notificationStreakSwitch.setOnCheckedChangeListener(new CompoundButton.OnCheckedChangeListener() {
//            @Override
//            public void onCheckedChanged(CompoundButton buttonView, boolean isChecked) {
//                streakPresenter.switchNotificationStreak(isChecked);
//                hideNotificationTime(!isChecked);
//            }
//        });
//        notificationIntervalValue.setText(notificationTimeValueString); //need to set for show default value, when user enable it
    }

    @Override
    public void hideNotificationTime(boolean needHide) {
        int visibility = needHide ? View.GONE : View.VISIBLE;
//        notificationTimeZoneInfo.setVisibility(visibility);
//        notificationIntervalTitle.setVisibility(visibility);
//        notificationIntervalValue.setVisibility(visibility);
    }

    @Override
    public void setNewTimeInterval(@NotNull String timePresentationString) {
//        notificationIntervalValue.setText(timePresentationString);
    }

    @Override
    public void onActivityResult(int requestCode, int resultCode, Intent data) {
        super.onActivityResult(requestCode, resultCode, data);
        if (requestCode == NOTIFICATION_INTERVAL_REQUEST_CODE) {
            if (resultCode == Activity.RESULT_OK) {
                int intervalCode = data.getIntExtra(TimeIntervalPickerDialogFragment.Companion.getResultIntervalCodeKey(), TimeIntervalUtil.INSTANCE.getDefaultTimeCode());
                streakPresenter.setStreakTime(intervalCode);
                analytic.reportEvent(Analytic.Streak.CHOOSE_INTERVAL_PROFILE, intervalCode + "");
            } else if (resultCode == Activity.RESULT_CANCELED) {
                analytic.reportEvent(Analytic.Streak.CHOOSE_INTERVAL_CANCELED_PROFILE);
            }
        }
    }


    @Override
    public void onCreateOptionsMenu(Menu menu, MenuInflater inflater) {
        super.onCreateOptionsMenu(menu, inflater);
        if (localUserViewModel != null) {
            inflater.inflate(R.menu.share_menu, menu);
        }
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        switch (item.getItemId()) {
            case R.id.menu_item_share:
                shareProfile();
                return true;
        }
        return false;
    }

    private void shareProfile() {
        if (localUserViewModel != null) {
            Intent intent = shareHelper.getIntentForProfileSharing(localUserViewModel);
            startActivity(intent);
        }
    }


}

File: app/src/main/java/org/stepic/droid/ui/fragments/ProfileFragment.kt
package org.stepic.droid.ui.fragments

import android.annotation.SuppressLint
import android.net.Uri
import android.os.Bundle
import android.view.LayoutInflater
import android.view.Menu
import android.view.MenuInflater
import android.view.MenuItem
import android.view.View
import android.view.ViewGroup
import android.widget.LinearLayout
import android.widget.TextView
import androidx.core.content.ContextCompat
import androidx.core.content.res.ResourcesCompat
import androidx.core.view.isInvisible
import androidx.core.view.isVisible
import androidx.recyclerview.widget.GridLayoutManager
import androidx.recyclerview.widget.LinearLayoutManager
import com.bumptech.glide.Glide
import com.bumptech.glide.load.engine.DiskCacheStrategy
import kotlinx.android.synthetic.main.empty_login.*
import kotlinx.android.synthetic.main.fragment_profile_new.*
import kotlinx.android.synthetic.main.latex_supportabe_enhanced_view.view.*
import kotlinx.android.synthetic.main.view_notification_interval_chooser.*
import org.stepic.droid.R
import org.stepic.droid.analytic.AmplitudeAnalytic
import org.stepic.droid.analytic.Analytic
import org.stepic.droid.base.App
import org.stepic.droid.base.FragmentBase
import org.stepic.droid.core.ProfilePresenter
import org.stepic.droid.core.presenters.StreakPresenter
import org.stepic.droid.core.presenters.contracts.NotificationTimeView
import org.stepic.droid.core.presenters.contracts.ProfileView
import org.stepic.droid.model.UserViewModel
import org.stepic.droid.ui.activities.MainFeedActivity
import org.stepic.droid.ui.activities.contracts.CloseButtonInToolbar
import org.stepic.droid.ui.adapters.ProfileSettingsAdapter
import org.stepic.droid.ui.dialogs.TimeIntervalPickerDialogFragment
import org.stepic.droid.ui.util.collapse
import org.stepic.droid.ui.util.expand
import org.stepic.droid.ui.util.initCenteredToolbar
import org.stepic.droid.util.AppConstants
import org.stepic.droid.util.DateTimeHelper
import org.stepic.droid.util.ProfileSettingsHelper
import org.stepic.droid.util.copyTextToClipboard
import org.stepic.droid.util.glide.GlideSvgRequestFactory
import org.stepic.droid.viewmodel.ProfileSettingsViewModel
import ru.nobird.android.view.base.ui.extension.argument
import timber.log.Timber
import java.util.ArrayList
import java.util.Date
import java.util.TimeZone
import javax.inject.Inject

class   ProfileFragment : FragmentBase(),
        ProfileView,
        NotificationTimeView,
        TimeIntervalPickerDialogFragment.Companion.Callback{

    companion object {
        private const val MAX_ACHIEVEMENTS_TO_DISPLAY = 6
        private const val DETAILED_INFO_CONTAINER_KEY = "detailedInfoContainerKey"

        fun newInstance(): ProfileFragment = newInstance(0)

        fun newInstance(userId: Long = 0) = ProfileFragment().apply {
            this.userId = userId
        }
    }

    @Inject
    lateinit var profilePresenter: ProfilePresenter

    @Inject
    lateinit var streakPresenter: StreakPresenter

    private var userId: Long by argument()
    private var localUserViewModel: UserViewModel? = null
    private var profileSettingsList: ArrayList<ProfileSettingsViewModel> = ArrayList()
    private var isShortInfoExpanded: Boolean = false

    private var achievementsToDisplay: Int = 0

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        analytic.reportEvent(Analytic.Profile.OPEN_SCREEN_OVERALL)
        setHasOptionsMenu(true)
        profileSettingsList.clear()
        profileSettingsList.addAll(ProfileSettingsHelper.getProfileSettings())
        if (userId == 0L) {
            userId = userPreferences.userId
        }
    }

    override fun injectComponent() {
        App
            .component()
            .profileComponentBuilder()
            .build()
            .inject(this)
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? =
            inflater.inflate(R.layout.fragment_profile_new, container, false)

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        achievementsToDisplay = resources.getInteger(R.integer.achievements_to_display)
        nullifyActivityBackground()
        super.onViewCreated(view, savedInstanceState)
        initToolbar()
        initTimezone()

        // app:fontFamily doesn't work on this view
        notificationStreakSwitch.typeface = ResourcesCompat.getFont(requireContext(), R.font.roboto_light)
        profileSettingsRecyclerView.layoutManager = LinearLayoutManager(context)
        profileSettingsRecyclerView.adapter = ProfileSettingsAdapter(requireActivity(), profileSettingsList, screenManager, this, analytic)
        profileSettingsRecyclerView.isNestedScrollingEnabled = false

        achievementsTilesContainer.layoutManager = GridLayoutManager(context, achievementsToDisplay)
        achievementsTilesContainer.isNestedScrollingEnabled = false
        initAchievementsPlaceholders()

        profilePresenter.attachView(this)
        streakPresenter.attachView(this)
        profilePresenter.initProfile(userId)

        profileImage.setOnClickListener { analytic.reportEvent(Analytic.Profile.CLICK_IMAGE) }
        val clickStreakValue = View.OnClickListener { analytic.reportEvent(Analytic.Profile.CLICK_STREAK_VALUE) }
        currentStreakValue.setOnClickListener(clickStreakValue)
        maxStreakValue.setOnClickListener(clickStreakValue)
        profileName.setOnClickListener { analytic.reportEvent(Analytic.Profile.CLICK_FULL_NAME) }

        notificationIntervalChooserContainer.setOnClickListener {
            analytic.reportEvent(Analytic.Interaction.CLICK_CHOOSE_NOTIFICATION_INTERVAL)
            val dialogFragment = TimeIntervalPickerDialogFragment.newInstance()
            if (!dialogFragment.isAdded) {
                dialogFragment.setTargetFragment(this@ProfileFragment, 0)
                dialogFragment.show(requireFragmentManager(), null)
            }
        }

        shortBioInfoContainer.setOnClickListener {
            changeStateOfUserInfo()
        }

        authAction.setOnClickListener {
            screenManager.showLaunchScreen(context, true, MainFeedActivity.PROFILE_INDEX)
        }

        shortBioSecondText.textView.textSize = 14f
        shortBioSecondText.textView.setLineSpacing(0f, 1.6f)

        viewAllAchievements.setOnClickListener { screenManager.showAchievementsList(context, localUserViewModel?.id ?: 0, localUserViewModel?.isMyProfile ?: false) }

        certificatesTitleContainer.setOnClickListener { screenManager.showCertificates(requireContext(), userId) }
    }

    override fun onDestroyView() {
        notificationStreakSwitch.setOnCheckedChangeListener(null)
        profileName.setOnClickListener(null)
        currentStreakValue.setOnClickListener(null)
        maxStreakValue.setOnClickListener(null)
        profileImage.setOnClickListener(null)
        notificationIntervalChooserContainer.setOnClickListener(null)
        streakPresenter.detachView(this)
        profilePresenter.detachView(this)
        shortBioInfoContainer.setOnClickListener(null)
        isShortInfoExpanded = detailedInfoContainer.visibility == View.VISIBLE
        super.onDestroyView()
        Timber.d("onDestroyView %s", this)
    }

    private fun changeStateOfUserInfo() {
        shortBioArrowImageView.changeState()
        val isExpanded = shortBioArrowImageView.isExpanded()
        isShortInfoExpanded = isExpanded
        if (isExpanded) {
            detailedInfoContainer.expand()
        } else {
            detailedInfoContainer.collapse()
        }
    }

    override fun onSaveInstanceState(outState: Bundle) {
        super.onSaveInstanceState(outState)
        outState.putBoolean(DETAILED_INFO_CONTAINER_KEY, isShortInfoExpanded)
    }

    override fun onViewStateRestored(savedInstanceState: Bundle?) {
        super.onViewStateRestored(savedInstanceState)
        savedInstanceState?.let {
            isShortInfoExpanded = it.getBoolean(DETAILED_INFO_CONTAINER_KEY)
            restoreVisibility(detailedInfoContainer, it, DETAILED_INFO_CONTAINER_KEY)
        }

    }

    private fun restoreVisibility(view: View, bundle: Bundle, bundleKey: String) {
        view.visibility = if (bundle.getBoolean(bundleKey)) {
            View.VISIBLE
        } else {
            View.GONE
        }
    }

    private fun initTimezone() {
        val timezone = TimeZone.getDefault()
        val nowDate = Date()
        val isDaylight = timezone.inDaylightTime(nowDate)
        val print = String.format("%s\n%s",
                DateTimeHelper.hourMinutesOfMidnightDiffWithUtc(timezone, nowDate),
                timezone.getDisplayName(isDaylight, TimeZone.LONG))
        notificationTimeZoneInfo.text = getString(R.string.streak_updated_timezone, print)
    }

    private fun initToolbar() {
        val needCloseButton = context is CloseButtonInToolbar
        initCenteredToolbar(R.string.profile_title, needCloseButton, getCloseIconDrawableRes())
    }

    private fun initAchievementsPlaceholders() {
        for (i in 0 until achievementsToDisplay) {
            val view = layoutInflater.inflate(R.layout.view_achievement_tile_placeholder, achievementsLoadingPlaceholder, false)
            view.layoutParams = (view.layoutParams as LinearLayout.LayoutParams).apply {
                weight = 1f
                width = 0
            }
            achievementsLoadingPlaceholder.addView(view)
        }
    }

//    override fun showAchievements(achievements: List<AchievementFlatItem>) {
//        (achievementsTilesContainer.adapter as BaseAchievementsAdapter).achievements = achievements.take(achievementsToDisplay)
//        achievementsLoadingPlaceholder.isVisible = false
//        achievementsLoadingError.isVisible = false
//        achievementsTilesContainer.isVisible = true
//        achievementsContainer.isVisible = true
//    }
//
//    override fun onAchievementsLoadingError() {
//        achievementsContainer.isVisible = true
//        achievementsLoadingPlaceholder.isVisible = false
//        achievementsLoadingError.isVisible = true
//        achievementsTilesContainer.isVisible = false
//    }
//
//    override fun onAchievementsLoading() {
//        achievementsContainer.isVisible = true
//        achievementsLoadingPlaceholder.isVisible = true
//        achievementsLoadingError.isVisible = false
//        achievementsTilesContainer.isVisible = false
//    }

    /**
     * This method is invoked only for My Profile
     */
    @SuppressLint("SetTextI18n")
    override fun streaksAreLoaded(currentStreak: Int, maxStreak: Int, haveSolvedToday: Boolean) {
        val suffixCurrent = resources.getQuantityString(R.plurals.day_number, currentStreak)
        val suffixMax = resources.getQuantityString(R.plurals.day_number, maxStreak)

        currentStreakValue.text = String.format("%d %s", currentStreak, suffixCurrent)
        maxStreakValue.text = String.format("%d %s", maxStreak, suffixMax)

        if (haveSolvedToday) {
            streakIndicator.setImageResource(R.drawable.ic_lightning)
        } else {
            streakIndicator.setImageResource(R.drawable.ic_lightning_inactive)
        }

        setStreakInfoVisibility(true)
    }

    override fun showLoadingAll() {
        contentRoot.visibility = View.GONE
        profileEmptyUser.visibility = View.GONE
        profileReportProblem.visibility = View.GONE
        profileNeedAuth.visibility = View.GONE
        profileLoadingView.visibility = View.VISIBLE
    }

    override fun showNameImageShortBio(userViewModel: UserViewModel) {
        profileNeedAuth.visibility = View.GONE
        profileEmptyUser.visibility = View.GONE
        profileReportProblem.visibility = View.GONE
        profileLoadingView.visibility = View.GONE
        contentRoot.visibility = View.VISIBLE

        localUserViewModel = userViewModel
        activity?.invalidateOptionsMenu()
        if (userViewModel.isMyProfile) {
            streakPresenter.tryShowNotificationSetting()

            //// TODO: 21.08.17 init here and do not spend resources for creating recycler on the another profiles
            profileSettingsRecyclerView.visibility = View.VISIBLE

            notificationIntervalChooserContainer.visibility = View.VISIBLE
            setupUserId()
        } else {
            //show user info expanded for strangers
            if (!shortBioArrowImageView.isExpanded()) {
                changeStateOfUserInfo()
            }
        }

        if (!userViewModel.isPrivate && !userViewModel.isOrganization) {
//            achievementsPresenter.showAchievementsForUser(userViewModel.id, MAX_ACHIEVEMENTS_TO_DISPLAY)
            certificatesTitleContainer.visibility = View.VISIBLE
        }

        mainInfoRoot.visibility = View.VISIBLE
        val nameArray = userViewModel
                .fullName
                .split("\\s+".toRegex())
                .dropLastWhile { it.isEmpty() }
                .toTypedArray()
        val builder = StringBuilder()
        for (nameArrayItem in nameArray) {
            if (builder.isNotEmpty()) {
                builder.append("\n")
            }
            builder.append(nameArrayItem)
        }


        profileName.text = builder.toString()
        val userPlaceholder = ContextCompat.getDrawable(requireContext(), R.drawable.general_placeholder)
        if (userViewModel.imageLink != null && userViewModel.imageLink.endsWith(AppConstants.SVG_EXTENSION)) {
            val svgRequestBuilder = GlideSvgRequestFactory.create(context, userPlaceholder)
            val uri = Uri.parse(userViewModel.imageLink)
            svgRequestBuilder
                    .diskCacheStrategy(DiskCacheStrategy.DATA)
                    .load(uri)
                    .into(profileImage)

        } else {
            Glide.with(requireContext())
                    .asBitmap()
                    .load(userViewModel.imageLink)
                    .placeholder(userPlaceholder)
                    .into(profileImage)
        }

        with(userViewModel) {
            shortBioInfoContainer.isVisible = shortBio.isNotBlank() || information.isNotBlank()
            shortBioSecondHeader.isVisible = shortBio.isNotBlank() && information.isNotBlank()
            when {
                shortBio.isBlank() && information.isNotBlank() ->
                    shortBioFirstHeader.setText(R.string.user_info) //show header with 'information'

                shortBio.isNotBlank() && information.isBlank() ->
                    shortBioFirstHeader.setText(R.string.short_bio)

                shortBio.isNotBlank() && information.isNotBlank() -> { //show general header and all info
                    shortBioFirstHeader.setText(R.string.short_bio_and_info)
                    shortBioSecondHeader.setText(R.string.user_info)
                }
            }

            if (shortBio.isBlank()) {
                shortBioFirstText.visibility = View.GONE
            } else {
                shortBioFirstText.text = shortBio.trim()
                shortBioFirstText.visibility = View.VISIBLE
            }

            if (information.isBlank()) {
                shortBioSecondText.visibility = View.GONE
            } else {
                shortBioSecondText.setPlainOrLaTeXTextWithCustomFontColored(
                        information, R.font.roboto_light, R.color.new_accent_color, false)
                shortBioSecondText.visibility = View.VISIBLE
            }
        }
    }

    override fun onInternetFailed() {
        profileLoadingView.visibility = View.GONE
        contentRoot.visibility = View.GONE
        profileEmptyUser.visibility = View.GONE
        profileNeedAuth.visibility = View.GONE
        profileReportProblem.visibility = View.VISIBLE
    }

    override fun onProfileNotFound() {
        profileLoadingView.visibility = View.GONE
        contentRoot.visibility = View.GONE
        profileReportProblem.visibility = View.GONE
        profileNeedAuth.visibility = View.GONE
        profileEmptyUser.visibility = View.VISIBLE
    }

    override fun onUserNotAuth() {
        profileLoadingView.visibility = View.GONE
        contentRoot.visibility = View.GONE
        profileReportProblem.visibility = View.GONE
        profileEmptyUser.visibility = View.GONE
        profileNeedAuth.visibility = View.VISIBLE
    }

    override fun showNotificationEnabledState(notificationEnabled: Boolean, notificationTimeValue: String) {
        notificationStreakSwitch.isChecked = notificationEnabled
        if (notificationStreakSwitch.visibility != View.VISIBLE) {
            notificationStreakSwitch.visibility = View.VISIBLE
        }
        if (notificationEnabled) {
            hideNotificationTime(false)
        } else {
            hideNotificationTime(true)
        }

        notificationStreakSwitch.setOnCheckedChangeListener { _, isChecked ->
            streakPresenter.switchNotificationStreak(isChecked)
            hideNotificationTime(!isChecked)
        }

        //need to set for show default value, when user enable it
        notificationIntervalTitle.text = resources.getString(R.string.notification_time, notificationTimeValue)
    }

    override fun hideNotificationTime(needHide: Boolean) {
        if (needHide) {
            notificationIntervalChooserContainer.collapse()
        } else {
            notificationIntervalChooserContainer.expand()
        }
    }

    override fun setNewTimeInterval(timePresentationString: String) {
        notificationIntervalTitle.text = resources.getString(R.string.notification_time, timePresentationString)
    }

    override fun onCreateOptionsMenu(menu: Menu, inflater: MenuInflater) {
        if (localUserViewModel != null) {
            inflater.inflate(R.menu.profile_menu, menu)

            menu.findItem(R.id.menu_item_edit)?.isVisible =
                localUserViewModel?.isMyProfile == true
        }
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean =
        when (item.itemId) {
            R.id.menu_item_share -> {
                shareProfile()
                true
            }
            R.id.menu_item_edit -> {
                analytic.reportAmplitudeEvent(AmplitudeAnalytic.ProfileEdit.SCREEN_OPENED)
                screenManager.showProfileEdit(context)
                true
            }
            else ->
                false
        }

    override fun onTimeIntervalPicked(chosenInterval: Int) {
        streakPresenter.setStreakTime(chosenInterval)
        analytic.reportEvent(Analytic.Streak.CHOOSE_INTERVAL_PROFILE, chosenInterval.toString() + "")
    }

    private fun shareProfile() {
        localUserViewModel?.let {
            val intent = shareHelper.getIntentForProfileSharing(it)
            startActivity(intent)
        }
    }

    private fun setStreakInfoVisibility(needShow: Boolean) {
        currentStreakSuffix.isInvisible = !needShow
        currentStreakValue.isInvisible = !needShow
        maxStreakSuffix.isInvisible = !needShow
        maxStreakValue.isInvisible = !needShow
        streakIndicator.isInvisible = !needShow
    }

    private fun setupUserId() {
        profileIdSeparator.visibility = View.VISIBLE
        profileId.visibility = View.VISIBLE
        profileId.text = getString(R.string.profile_user_id, userId)
        profileId.setOnLongClickListener {
            val textToCopy = (it as TextView).text.toString()
            requireContext().copyTextToClipboard(textToCopy = textToCopy, toastMessage = getString(R.string.copied_to_clipboard_toast))
            true
        }
    }
}
